File: leetcode_by_cpp/145.二叉树的后序遍历.cpp
/*
 * @lc app=leetcode.cn id=145 lang=cpp
 *
 * [145] 二叉树的后序遍历
 *
 * https://leetcode-cn.com/problems/binary-tree-postorder-traversal/description/
 *
 * algorithms
 * Medium (72.31%)
 * Likes:    392
 * Dislikes: 0
 * Total Accepted:    118.3K
 * Total Submissions: 163.6K
 * Testcase Example:  '[1,null,2,3]'
 *
 * 给定一个二叉树，返回它的 后序 遍历。
 * 
 * 示例:
 * 
 * 输入: [1,null,2,3]  
 * ⁠  1
 * ⁠   \
 * ⁠    2
 * ⁠   /
 * ⁠  3 
 * 
 * 输出: [3,2,1]
 * 
 * 进阶: 递归算法很简单，你可以通过迭代算法完成吗？
 * 
 */
#include <vector>
#include <stack>
#include <algorithm>
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
private:
    vector<int> res;

public:
    void postOrder(TreeNode *root)
    {
        if (!root)
            return;
        postOrder(root->left);
        postOrder(root->right);
        res.push_back(root->val);
    }
    vector<int> postorderTraversal(TreeNode *root)
    {
        /* if(!root)
            return res;
        postOrder(root);
        return res; */
        stack<TreeNode *> s;
        s.push(root);
        while (!s.empty())
        {
            TreeNode *cur = s.top();
            s.pop();
            if (cur)
                res.push_back(cur->val);
            else
                continue;
            s.push(cur->left);
            s.push(cur->right);
        }
        reverse(res.begin(),res.end());
        return res;
    }
};
// @lc code=end

File: leetcode_by_rust/src/bin/145.二叉树的后序遍历.rs
/*
 * @lc app=leetcode.cn id=145 lang=rust
 *
 * [145] 二叉树的后序遍历
 *
 * https://leetcode.cn/problems/binary-tree-postorder-traversal/description/
 *
 * algorithms
 * Easy (76.40%)
 * Likes:    1157
 * Dislikes: 0
 * Total Accepted:    729.6K
 * Total Submissions: 953.5K
 * Testcase Example:  '[1,null,2,3]'
 *
 * 给你一棵二叉树的根节点 root ，返回其节点值的 后序遍历 。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：root = [1,null,2,3]
 * 输出：[3,2,1]
 *
 *
 * 示例 2：
 *
 *
 * 输入：root = []
 * 输出：[]
 *
 *
 * 示例 3：
 *
 *
 * 输入：root = [1]
 * 输出：[1]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 树中节点的数目在范围 [0, 100] 内
 * -100 <= Node.val <= 100
 *
 *
 *
 *
 * 进阶：递归算法很简单，你可以通过迭代算法完成吗？
 *
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn postorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut res = vec![];
        let mut stack = VecDeque::new();
        stack.push_back(root.map(|x| (x, false)));
        while let Some(x) = stack.pop_back() {
            match x {
                Some((node, true)) => res.push(node.borrow().val),
                Some((node, false)) => {
                    stack.push_back(Some((node.clone(), true)));
                    stack.push_back(node.borrow().right.clone().map(|x| (x, false)));
                    stack.push_back(node.borrow().left.clone().map(|x| (x, false)));
                }
                _ => (),
            }
        }
        res
    }
}
// @lc code=end
